### PR TITLE
Lift code loading to Container from ContainerContext

### PR DESCRIPTION
**Pull request by ChumpChief**
This change lifts code loading and `satisfies` checking up to the `Container` from the `ContainerContext`.  There are a few motivations for this:
1. Consolidating the quorum/codeDetails/codeLoader/comparers/etc. lets us avoid over-exposing them among classes.  ContainerContext can relieve itself of knowing about code loading and details.  ContainerContext gets a simpler construction and can drop a couple public members.
2. Container can load the runtime factory before instantiating the context.  This allows injecting an already-realized factory, so ContainerContext doesn't have to think about async retrieval of the factory.
3. Ultimately it's my opinion that code loading should be externalized from the Container as well (should be completely in the hands of the host who passes the code loader today).  This gets us a little closer to popping it out, though I avoid changes to API boundaries for now.

A few key surprising discoveries that explain how this is a safe change:
1. Although it looks like the factory is lazy-loaded (via a `LazyPromise`), we in fact immediately await the promise in all cases (ContainerContext.createOrLoad immediately awaits the instantiateRuntime call before even returning a reference).  So we don't lose any laziness with this change.
2. The LegacyCodeLoader path has apparently been broken for about a year (looks like a bad merge).  If loadCodeModule did in fact ever return a module with no "module" member, then it would immediately blow up in getRuntimeFactory as PackageNotFactoryError.  So we don't need to maintain that path anymore since it's been broken for a year - telemetry supports this.
3. The tests aren't actually testing the flow, so I just remove them here.
    * The "legacy loader" test isn't actually testing for using a legacy loader anymore (otherwise it would have caught the issue in (2)).  Looks like this got updated to the "new thing" without realizing that it was intentionally "doing the old thing".
    * The "without details" test mocks returning a module without details during load (which I don't think is a real path anyway), but the details don't come into play during load anyway so this isn't interesting.  No details would be interesting during satisfies, but the satisfies call is mocked here so it's not actually testing the implementation - just testing the mock.
    * The "with details" test is similar to the "without details" - over-mocked so doesn't end up testing anything interesting.

### CODE REVIEW

File: packages/loader/container-loader/src/container.ts [hunk 1]
```diff
@@ -114,6 +118,8 @@ const detachedContainerRefSeqNumber = 0;
 const dirtyContainerEvent = "dirty";
 const savedContainerEvent = "saved";
 
+const PackageNotFactoryError = "Code package does not implement IRuntimeFactory";
```
Comment (scottn12) [inline packages/loader/container-loader/src/container.ts:121]: nit

```suggestion
const packageNotFactoryError = "Code package does not implement IRuntimeFactory";
```

File: packages/loader/container-loader/src/test/containerContext.spec.ts
Comment (scottn12) [inline packages/loader/container-loader/src/test/containerContext.spec.ts:1]: I know you mentioned these tests aren't actually useful, but could we include creating new useful tests as part of this PR? Even if this functionality is implicitly tested in other tests, I would argue it is still useful to have explicit tests for the individual functionality to make it easier to identify if it ever breaks.

**Review by scottn12 — COMMENTED** [review]
Looks good, thanks Matt! My only concern is related to testing and I left a separate comment for that :)

File: packages/loader/container-loader/src/test/containerContext.spec.ts
Comment (ChumpChief) [reply packages/loader/container-loader/src/test/containerContext.spec.ts:1]: It's a good idea, but I think a larger work item -- we don't actually have any unit tests for `Container` currently, so I think there's some work to figure out a reasonable way to mock and test it.  container.spec.ts is another over-mocked test (doesn't actually instantiate any real class we ship at all).

The end to end tests go through Loader to get an IContainer, but we also don't have any unit tests for Loader.

I can open a backlog ticket for figuring a unit test approach for the package more generally maybe?

**Review by scottn12 — APPROVED** [review]

**Comment by ChumpChief** [comment]
Looking at the test failures (in fluid-runner) now.  It appears these tests should have always failed, but were sneaking by because of the async module resolution (they "passed" before hitting the error), so now that we await less the errors are surfacing.  Tossing a few `await Promise.resolve().then(() => {});` statements in `satisfies` makes them "pass" again.

For example, odspSnapshot2.json includes a propose op.  The sampleCodeLoader is not an IFluidCodeDetailsComparer, so this should result in force-close of the container and probably some sort of failure (both before and after this PR).  Will take a bit of time to figure out what the right test fix is.

**Comment by kian-thompson** [comment]
> Looking at the test failures (in fluid-runner) now. It appears these tests should have always failed, but were sneaking by because of the async module resolution (they "passed" before hitting the error), so now that we await less the errors are surfacing. Tossing a few `await Promise.resolve().then(() => {});` statements in `satisfies` makes them "pass" again.
> 
> For example, odspSnapshot2.json includes a propose op. The sampleCodeLoader is not an IFluidCodeDetailsComparer, so this should result in force-close of the container and probably some sort of failure (both before and after this PR). Will take a bit of time to figure out what the right test fix is.

You're correct that the containers have been closing silently in the background while we have these sample js bundles doing no actual work on the container. Is there any kind of exception being thrown now that is causing these tests to fail? Before, the container would close, but no actual exception was thrown to prevent the test from continuing.

**Comment by ChumpChief** [comment]
> You're correct that the containers have been closing silently in the background while we have these sample js bundles doing no actual work on the container. Is there any kind of exception being thrown now that is causing these tests to fail? Before, the container would close, but no actual exception was thrown to prevent the test from continuing.

Looking at exportFile's "Export using snapshot [odspSnapshot2.json]" specifically:  both with and without this change, a "Existing context does not satisfy incoming proposal" error is thrown (which is correct - the code loader doesn't implement comparison, so it is correct to throw this error as the test is currently written).

The only thing that's different is the timing of that error being thrown.  With this change, there are fewer `await`s before the error is realized (because of the removal of the LazyPromise), so it happens sooner.  Hypothesis is: with the delay, the test was able to [forcibly close() the container here](https://github.com/microsoft/FluidFramework/blob/main/packages/tools/fluid-runner/src/exportFile.ts#L126) before it processed all the ops and encountered the error.  Now that the delay is gone the ops get processed and the error becomes apparent.  Still probing to confirm.

I would like to understand what this test is trying to verify though - was it correct to declare success without processing all the ops?  Why is it writing the word "result" into a file, etc.

**Comment by ChumpChief** [comment]
Latest approach implements a naive comparison implementation for the fluid-runner's code loader.  Chatted with Kian and it sounds like there is followup work planned to improve these tests, so this should tide us over until that work is done.

**Pull request merged**